build: run linter before running tests

joyeecheung opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
build, test

Fixes: https://github.com/nodejs/node/issues/16283

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10807/

joyeecheung commented:
cc @nodejs/testing

Trott commented:
A bit of probably-relevant back-story: I made this change over a year ago in https://github.com/nodejs/node/commit/d9f7a597e4d2f5cefda898f8b1223e3fcfb7a07c and people hated it enough that I felt compelled to revert it in https://github.com/nodejs/node/pull/5602.

I like it, though. +1 from me!

Trott commented:
Another option, if people do object to this, might be to make it somehow configurable or have a task separate from `make test` that has this ordering.

gibfahn commented:
Great idea! Should make things easier for new contributors too (number of people running `make test` and finding linter errors at the end at code-and-learn was large).

joyeecheung commented:
Looking at https://github.com/nodejs/node/pull/5470 cc @bnoordhuis @Fishrock123 

I am thinking maybe we can have two targets

* `make test-only` for running tests (better for people who are familiar with our process)
* `make test` for running lint + tests (better for new contributors)

Doesn't have to be these two, any suggestions are welcomed. (I wouldn't even object to `make t` because hey, less typing)

refack commented:
> LGTM. Maybe the thing to do to make everyone happy is put the linter into a test file so it can run in parallel with other tests. :-D I'm just joking. OR AM I?!?!

I like the "linter as test" idea (or some other way to make `lint` not stop the rest of the `make`), since it will allow to get both outputs with a single run.

gibfahn commented:
>I am thinking maybe we can have two targets
>
>* `make test-only` for running tests (better for people who are familiar with our process)
>* `make test` for running lint + tests (better for new contributors)

Something like this seems like it would resolve @bnoordhuis 's objection to the previous PR, while making new contributor's lives easier.

>(I wouldn't even object to make t because hey, less typing)

I like `make t`, it's not obvious, but then this isn't something people getting started are likely to need.

jasnell commented:
Another possible color for the bikeshed: A `./configure --lint-first` type of option may be worthwhile... that is, some configuration option that would instruct the `Makefile` to run linting first.

joyeecheung commented:
@jasnell That works as well but I think the default should be lint-first instead? Also this could require some hack because the main Makefile is not actually generated...

mhdawson commented:
How about "test-no-lint" for the target that does everything except for the linter ?  But no strong feeling, just another suggestion to consider.

joyeecheung commented:
Rebased and implemented `make t` for people who just want a quick check. @apapirovski @refack @Trott @lpinca @BridgeAR @mhdawson @gibfahn Can people who approved before take another look please? Thanks!

Also @bnoordhuis do you still object to this idea? Although we can always revert if we hate it ¯\_(ツ)_/¯ 

CI: https://ci.nodejs.org/job/node-test-pull-request/10845/

gibfahn commented on Makefile:
```diff
@@ -209,15 +209,26 @@ test: all
 endif
 
+# For a quick test, does not run linter or build doc
+t: all
```
Does `t` need to be added to `.PHONY`?

richardlau commented on Makefile:
```diff
@@ -209,15 +209,26 @@ test: all
 endif
 
+# For a quick test, does not run linter or build doc
+t: all
```
`t` is a terrible name for a make target (that what it does has to be described in a comment is a good indicator that it is badly named). We're already at 40+ phony targets -- It should be in our interest that the target names be reasonably self descriptive.

gibfahn replied:
But this isn't something most people will want to use. It looks like a shortcut, which is what it is.

Fishrock123 replied:
I would prefer `test-only`, personally.

joyeecheung commented:
@richardlau @tniessen @mhdawson Yeah...took the name because it's meant to be a shortcut, couldn't think of anything better at that point. How about `make check`

gibfahn commented:
>How about make check?

Already exists, and is used by people unfortunately.

https://github.com/nodejs/node/blob/bf1bacef6bc49fa39a272ca63b7635de3e5e7f08/Makefile#L122

joyeecheung commented:
@gibfahn uh-oh...`make try`? (still short)

tniessen commented:
@silverwind I liked your original idea better (`make test` vs `make test-only`) simply because I prefer self-explanatory names, but feel free to stick to something short if the majority prefers that :)

apapirovski commented:
I like the longer name in theory but in practice it might not get used... Personally, I would use `make t` solely because it's less typing and because I have the linter running in `atom` anyway. I would probably not use `make test-only` because it's longer than `make test` (and I would constantly forget if there's a dash in it or not).

But I'm fine with whatever the rest decide on, just wanted to explain why I backed `make t` above.

Fishrock123 commented on Makefile:
```diff
@@ -209,15 +209,26 @@ test: all
 	$(MAKE) build-addons
 	$(MAKE) build-addons-napi
 	$(MAKE) doc
+	$(MAKE) lint
```
IMO this should be at the top, infront of any of the native builds then?

joyeecheung commented:
OK I am just going to do a poll here. For the lint + test rule:

`make t`: 👍
`make try`: 🎉
`make test-only`: ❤️

@Fishrock123 @richardlau @gibfahn @apapirovski @refack @Trott @lpinca @tniessen @BridgeAR @mhdawson

Can people previously approved this give a quick vote please? Thanks!

joyeecheung commented:
One week has passed after the poll, `test-only` winned.

Rebased and changed `t` to `test-only`. New CI: https://ci.nodejs.org/job/node-test-pull-request/11045/

I plan to land this on Sunday if there are no strong -1s.

refack commented on Makefile:
```diff
@@ -213,15 +213,26 @@ test: all
 	$(MAKE) build-addons
 	$(MAKE) build-addons-napi
 	$(MAKE) doc-only
+	$(MAKE) lint
```
Non Blocking Suggestion: run `lint` before `all`

joyeecheung replied:
My plan is to make `make lint` independent of the binary build, if we do that `lint` will look for a global `node` which might not exist.

refack replied:
Ack. (I'm used to run it as `make lint PYTHON=python NODE=node`)

refack replied:
another future TODO: `make lint-changes` that queries git for the changed files since...

richardlau commented:
`test-only` should be added to thevlist of phony targets.

joyeecheung commented:
@richardlau Added, thanks for pointing that out! 

New CI: https://ci.nodejs.org/job/node-test-pull-request/11051/

joyeecheung commented:
Landed in 287545997288b038c0e8a84f44d338e4d1acc384, thanks!

MylesBorins commented:
This doesn't land cleanly on v6.x-staging can someone manually backport?
